### PR TITLE
Add new function insert_and_find for static_map

### LINKED ISSUE #200: [BUG] `static_map::device_mutable_view::insert` should return iterator to inserted / existing element
**Describe the bug**
Commit https://github.com/NVIDIA/cuCollections/commit/0145753a5531aade683326981d98196aa8a56bac changed `insert` to only return a `bool`.

**Expected behavior**
The long form of the documentation still says:
https://github.com/NVIDIA/cuCollections/blob/e91e9dabc9d2855a70446de9b3f9ee6208220be5/include/cuco/static_map.cuh#L925-L927
This is required to implement insert-or-update access; right now a kernel would need both a (non-mutable) `device_view` to call `find` and a `device_mutable_view` to call `insert`. I'm not even sure if using the former in parallel with insertions is even safe?

**Additional context**
`std::unordered_map` provides the same semantics of returning an iterator with `insert`.

### PR DESCRIPTION
**Pull request by hahnjo**
It inserts a pair as the existing function insert does, but also returns an iterator (and makes sure it is valid).

Closes #200

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

**Comment by sleeepyjack** [comment]
ok to test

File: include/cuco/detail/static_map.inl
Comment (sleeepyjack) [inline include/cuco/detail/static_map.inl:428]: We may want a more descriptive strong type here, e.g., `cuco::insert_result`. A plain `thrust::pair`could mean anything.

File: include/cuco/detail/static_map.inl
Comment (sleeepyjack) [inline include/cuco/detail/static_map.inl]: This always selects the `cas_dependent_write` strategy.

Maybe I missed the relevant section in our discussion but why do we eliminate the `back_to_back_cas` entirely for this function?

File: include/cuco/static_map.cuh
Comment (sleeepyjack) [inline include/cuco/static_map.cuh]: We now return `thrust::pair<...>` or better `cuco::insert_result`

File: tests/static_map/fetch_insert_test.cu
Comment (sleeepyjack) [inline tests/static_map/fetch_insert_test.cu]: I would just use an additional template param `typename FilterView` and let the compiler figure out the correct type.

File: include/cuco/detail/static_map.inl
Comment (sleeepyjack) [inline include/cuco/detail/static_map.inl:461]: I'm not entirely convinced that this is safe if we take deletions into account.
The current spin loop relies on the value slot only being able to transition from empty->filled but a deletion triggers filled->empty. So maybe we need to disable this functionality if deletions are enabled.

File: include/cuco/detail/static_map.inl
Comment (jrhemstad) [inline include/cuco/detail/static_map.inl:458]: I see what you mean now. 

The idea is to prevent another thread from returning an iterator to a slot in a partial state by spinning on the value. 

We'd need to think about this very carefully to make sure there aren't any subtle memory ordering or ABA issues with this approach. I'll make @sleeepyjack do that work for me ;) 

This approach also can't work on anything below Volta as it requires independent thread scheduling, so we'd need to guard it based on the arch (which is fine).

File: include/cuco/detail/static_map.inl
Comment (hahnjo) [reply include/cuco/detail/static_map.inl]: It's alluded to by the commit right above: We cannot guarantee a valid iterator with the `back_to_back_cas`. It's also been discussed here: https://github.com/NVIDIA/cuCollections/issues/200#issuecomment-1205392853

In short, the problematic case is if two threads want to insert the same key and one gets the key, the other gets the value. Now if a third thread comes along, it will see the key it wants and a non-empty value. However, the second thread that got the value, will first reset it to the empty sentinel first to signal that the first thread can write its value.

File: include/cuco/detail/static_map.inl
Comment (jrhemstad) [reply include/cuco/detail/static_map.inl:428]: ~tbh, I'd love to return a `thrust::optional<iterator>` here.~

Ignore me, this isn't `find`. Remind me not to review PRs first thing in the morning :sweat_smile:

File: include/cuco/detail/static_map.inl
Comment (hahnjo) [reply include/cuco/detail/static_map.inl:461]: I *think* this is fine because it shouldn't be possible to write legitimate code that concurrently deletes a key and modifies its value. Even if this was made to work here, user code still wouldn't be allowed to dereference an iterator because another thread might have erased it. Probably a good idea to document this; also users must not mix `insert` and `fetch_insert` because the former could still fall back to `back_to_back_cas`.

File: include/cuco/detail/static_map.inl
Comment (hahnjo) [reply include/cuco/detail/static_map.inl:458]: > This approach also can't work on anything below Volta as it requires independent thread scheduling, so we'd need to guard it based on the arch (which is fine).

True, the spin loop needs the forwards guarantee that the inserting thread eventually writes the value...

File: include/cuco/detail/static_map.inl
Comment (sleeepyjack) [reply include/cuco/detail/static_map.inl:461]: Ah, yes, you're right! The constrains that apply to concurrent `ìnsert`+`erase` also apply to `fetch_insert`+`erase`.

> also users must not mix insert and fetch_insert because the former could still fall back to back_to_back_cas.

If we disable `fetch_insert` for pre-Volta architectures, then there should be no issue in using `insert`+`fetch_insert` concurrently. A simple `#if (CUCO_CUDA_MINIMUM_ARCH >=700)` should do the trick. BTW this macro is introduced by #201 so you might need to cherry-pick those commits first.

Comment (hahnjo) [reply include/cuco/detail/static_map.inl:461]: > > also users must not mix insert and fetch_insert because the former could still fall back to back_to_back_cas.
> 
> If we disable `fetch_insert` for pre-Volta architectures, then there should be no issue in using `insert`+`fetch_insert` concurrently. A simple `#if (CUCO_CUDA_MINIMUM_ARCH >=700)` should do the trick. BTW this macro is introduced by #201 so you might need to cherry-pick those commits first.

Unfortunately, it's exactly the other way around: pre-Volta uses `cas_dependent_write` and we likely can't support `fetch_insert` anyway; Volta+ uses `back_to_back_cas` for `insert` (for performance reasons, but I personally don't see much of a difference on my Turing card), which you must not mix with `fetch_insert`.

Comment (sleeepyjack) [reply include/cuco/detail/static_map.inl:461]: Dang, sorry for my confusion :D Well, in that case, I agree that we should add a section about this in the documentation.

**Comment by PointKernel** [comment]
ok to test

File: include/cuco/detail/static_map.inl
Comment (hahnjo) [reply include/cuco/detail/static_map.inl]: I've expanded the comment.

**Comment by hahnjo** [comment]
(For this repository I should push new commits to address review comments, right? Otherwise I'd rewrite the history to squash the commits, but I read somewhere that I should not force-push; not sure where...)

File: include/cuco/detail/static_map.inl
Comment (PointKernel) [reply include/cuco/detail/static_map.inl:458]: Imagine there are 3 threads: T0, T1 and T2 inserting `{5, 1}`, `{5, 2}` and `{5, 3}`
T0 wins key CAS but not value CAS
T1 wins value CAS but not key CAS
So the slot content becomes `{5, 2}`
When T2 comes in, it finds the slot is not available and the value is not sentinel. So it returns a pointer to `{5, 2}` while the expected would be `{5, 1}`.

Did I miss anything?

Comment (hahnjo) [reply include/cuco/detail/static_map.inl:458]: Yes, that's exactly the reason why we cannot use `back_to_back_cas` for implementing `fetch_insert`.

**Comment by PointKernel** [comment]
> For this repository I should push new commits to address review comments, right? 

Yeah, that's fine.

**Comment by sleeepyjack** [comment]
> (For this repository I should push new commits to address review comments, right? Otherwise I'd rewrite the history to squash the commits, but I read somewhere that I should not force-push; not sure where...)

I haven't noticed such a guideline and I regularly rebase my PRs in order to have a clean history.

File: include/cuco/detail/static_map.inl
Comment (PointKernel) [reply include/cuco/detail/static_map.inl:428]: I guess @sleeepyjack's idea is to create a `cuco::insert_result` struct with meaningful member names to explicitly denote the pair-like return type. Something like:
```cpp
struct fetch_insert_result {
  Iterator iter;
  bool is_successful;
};
```
I would preder `thrust::pair<iterator, bool>` since it's std-canonical (see [here](https://en.cppreference.com/w/cpp/container/unordered_map/insert) and [here](https://en.cppreference.com/w/cpp/container/unordered_set/insert)).

File: include/cuco/static_map.cuh
Comment (PointKernel) [inline include/cuco/static_map.cuh]: @hahnjo Can you please add a `static_assert` in the implementation instead of guarding the public API with an arch-dependent macro?
e.g.
```cpp
#if __CUDA_ARCH__ < 700
  static_assert(cuco::detail::is_packable<value_type>(), "fetch insert is not supported for unpackable data on pre-Volta GPUs.");
#endif 
```

File: include/cuco/static_map.cuh
Comment (PointKernel) [inline include/cuco/static_map.cuh]:
```suggestion
     * For the same reason, the function must not be called concurrently
```

File: include/cuco/static_map.cuh
Comment (PointKernel) [inline include/cuco/static_map.cuh:955]:
```suggestion
     * @tparam KeyEqual Binary callable type
     
```
Blank line between `tparam` and `param`

File: tests/static_map/fetch_insert_test.cu
Comment (PointKernel) [inline tests/static_map/fetch_insert_test.cu]:
```suggestion
    auto [iter, is_successful] = v.fetch_insert(thrust::make_pair(i, 1));
    // for debugging...
    // if (iter->second < 0) {
    //   asm("trap;");
    // }
    if (! is_successful) { iter->second += 1; }
```
Use structured binding for readability

**Review by PointKernel — CHANGES_REQUESTED** [review]
Looks great!

Just some nitpicks. As for the unpackable handling, I don't see any issues with `cas_dependent_write`.

File: include/cuco/detail/static_map.inl
Comment (jrhemstad) [reply include/cuco/detail/static_map.inl:458]: Come to think of it, I recall @nicolas-iskos experimented with something like this back when he originally discovered the original race condition that required us to stop returning an iterator. Nico, do you recall why we didn't move forward with that idea?

File: include/cuco/detail/static_map.inl
Comment (hahnjo) [reply include/cuco/detail/static_map.inl:428]: Yes, this is the idea behind returning a `pair`, that the signature conforms to `std::unordered_map` in the STL.

File: include/cuco/static_map.cuh
Comment (hahnjo) [reply include/cuco/static_map.cuh]: Done. Could you give this a try pre-Volta to make sure the test passes there? I don't have easy access and the CI seems to be V100s...

File: include/cuco/detail/static_map.inl
Comment (sleeepyjack) [reply include/cuco/detail/static_map.inl:428]: Yes, I find this design decision in the standard questionable from the outset. However, following the standard should minimize confusion. So let's do it this way.

File: include/cuco/static_map.cuh
Comment (sleeepyjack) [reply include/cuco/static_map.cuh]: I just ran the tests on a Pascal card and everything seems to be fine.

**Comment by sleeepyjack** [comment]
@PointKernel @jrhemstad Do we also want a host-side API for this?

**Comment by jrhemstad** [comment]
> @PointKernel @jrhemstad Do we also want a host-side API for this?

I'm not sure of the value of returning iterators in a bulk-API. 

I suspect the `insert` overload I described here https://github.com/NVIDIA/cuCollections/issues/111#issuecomment-967248823 would satisfy any usecase that may want a bulk API for `fetch_insert`.

**Comment by PointKernel** [comment]
> I suspect the insert overload I described here #111 (comment) would satisfy

Agreed. Adding host API could be a separate PR.

File: include/cuco/static_map.cuh
Comment (jrhemstad) [inline include/cuco/static_map.cuh]: I'd avoid an alias here tbh. I think it's clearer to just be explicit about the return type on the function declaration.

Comment (hahnjo) [reply include/cuco/static_map.cuh]: @sleeepyjack asked for a more descriptive type name in https://github.com/NVIDIA/cuCollections/pull/202#discussion_r938754571. As discussed further, I may have misunderstood the request, so happy to revert back...

Comment (PointKernel) [reply include/cuco/static_map.cuh]: Please get rid of the alias and directly use `thrust::pair<iterator, bool>`.

File: include/cuco/static_map.cuh
Comment (PointKernel) [inline include/cuco/static_map.cuh:953]:
```suggestion
     *
     * Concurrent `fetch_insert`/`find` is supported if `supports_concurrent_insert_find()` returns true.
     *  
```

Comment (hahnjo) [reply include/cuco/static_map.cuh:953]: I've added the note to the previous paragraph.

**Review by PointKernel — APPROVED** [review]
@hahnjo Great work! 

Thanks for the contribution!

File: include/cuco/detail/static_map.inl
Comment (Nicolas-Iskos) [reply include/cuco/detail/static_map.inl:458]: IIRC, I implemented something very similar to this to attempt to support concurrent insert/find. I think at that time, it slowed down both `insert` and `find`. There may have also been another reason, but I can't recall it at the moment. Perhaps one way to measure the perf impact would be to use `packed_cas` for 4B/4B pairs and compare throughput to the `cas_dependent_write` + spin loop?

File: include/cuco/detail/static_map.inl
Comment (Nicolas-Iskos) [inline include/cuco/detail/static_map.inl:427]: Are there plans to add a CG version of the device-side function? What about a bulk host-side API?

Comment (jrhemstad) [reply include/cuco/detail/static_map.inl:427]: We discussed a bulk-API here: https://github.com/NVIDIA/cuCollections/pull/202#issuecomment-1208131266

A CG version would be nice, but I wouldn't consider it required for this PR. It can always be added later.

File: include/cuco/detail/static_map.inl
Comment (jrhemstad) [reply include/cuco/detail/static_map.inl:458]: > Perhaps one way to measure the perf impact would be to use packed_cas for 4B/4B pairs and compare throughput to the cas_dependent_write + spin loop?

It would be interesting to see what the perf difference is with this implementation. 

@hahnjo would you be interested in making a benchmark for this new API? It would be nice, but I also don't want to hold this PR back unnecessarily if you'd prefer to just get it merged and we can revisit benchmarking in a future PR.

Comment (hahnjo) [reply include/cuco/detail/static_map.inl:458]: > Perhaps one way to measure the perf impact would be to use packed_cas for 4B/4B pairs and compare throughput to the cas_dependent_write + spin loop?

Note that `fetch_insert` will still use `packed_cas` if it can, and in that case the performance should be identical to `insert`.

> @hahnjo would you be interested in making a benchmark for this new API?

I tried to do some benchmarking last week, but IIRC they all use the host APIs which we don't have yet for this new function. So I temporarily tried replacing the functionality of the `insert` host API with the new `fetch_insert`, and I got very weird results: For a start, `cas_dependent_write` seems to be faster than `back_to_back_cas` on my RTX 2070 SUPER. Then having the additional spin loops for `fetch_insert` wasn't much slower.

> It would be nice, but I also don't want to hold this PR back unnecessarily if you'd prefer to just get it merged and we can revisit benchmarking in a future PR.

Yes, I'd prefer to get this merged as-is. If somebody has a good idea how to benchmark this, I'm happy to provide my thoughts on a follow-up PR.

File: include/cuco/detail/static_map.inl
Comment (jrhemstad) [inline include/cuco/detail/static_map.inl:483]: I just realized that this means there are situations where it is unsafe to use `insert` and `fetch_insert` concurrently.

e.g., if `insert` uses `back_to_back_cas` concurrently with a `fetch_insert` using `cas_dependent_write` and the `fetch_insert` wins the key CAS and the `insert` succeeds in the value CAS then `fetch_insert` may store to the payload only for it to be overwritten by the `insert` restoring the `empty_value_sentinel`.

This isn't a showstopper for this feature, but at the very least we should document that `insert` and `fetch_insert` shouldn't be used concurrently.

Comment (jrhemstad) [reply include/cuco/detail/static_map.inl:483]: Nevermind, I see you already documented it :wink:

File: include/cuco/static_map.cuh
Comment (jrhemstad) [inline include/cuco/static_map.cuh]: "packable" types is an implementation detail. It's sufficient to just say that the performance may be worse and to prefer the other flavor of `insert` when possible. 


```suggestion
     * Note: In order to guarantee the validity of the returned iterator, `fetch_insert` may be less efficient than `insert` in some situations. Prefer using `insert` unless the returned iterator is required.
     *
     * Note: `fetch_insert` may only be used concurrently with `insert`, `find`, and `erase` when `supports_concurrent_insert_find()` returns true. 
```

File: include/cuco/detail/static_map.inl
Comment (jrhemstad) [reply include/cuco/detail/static_map.inl:458]: >  I got very weird results: For a start, cas_dependent_write seems to be faster than back_to_back_cas on my RTX 2070 SUPER

We've never robustly tested on which platforms `back_to_back_cas` is faster than `cas_dependent_write`. I think we've only really checked V100 and A100, so it's very possible that the latter could be faster on some architectures or with GDDR instead of HBM. 

> Yes, I'd prefer to get this merged as-is. 

Sounds good.

File: tests/static_map/fetch_insert_test.cu
Comment (sleeepyjack) [inline tests/static_map/fetch_insert_test.cu]:
```suggestion
 * Copyright (c) 2022, NVIDIA CORPORATION.
 * Copyright (c) 2022, Jonas Hahnfeld, CERN.
```

Comment (hahnjo) [reply tests/static_map/fetch_insert_test.cu]: I don't see why NVIDIA would have any copyright in this file; I wrote this test from scratch.

Comment (sleeepyjack) [reply tests/static_map/fetch_insert_test.cu]: You're right. I had to check our guidelines on this. Resolved.

**Comment by jrhemstad** [comment]
I think the implementation here is good to go, but I don't think `fetch_insert` is the right name.

The `fetch_op` semantics of `atomic(_ref)` is: "Do `op` and return the old value". So it's fetch _then_ op. 

The semantics of this function are "Do the insert and return the result". So this is op _then_ fetch. It reverses the order. 

Here's some options for what I think we could do instead:

1. Rename `fetch_insert` to `insert_and_fetch`.
   - This makes the order of operations clear 
2. Rename the `bool insert(...)` overload to `bool try_insert(...)` and `pair<iterator, bool> fetch_insert(...)` to `pair<iterator, bool> insert()`
   -  This is appealing to keep `insert` more symmetrical with the `std::` data structures, but would break the existing `insert` API
   - A breaking change isn't the end of the world here, but would be a bit annoying

I'd opt for option 1 just because it's easier. I think we'll consider option 2 as part of the redesign we're doing.

**Comment by hahnjo** [comment]
@jrhemstad I've renamed the function to `insert_and_find`, which I think is even clearer: the `fetch` operation is clearly defined for atomics, but for maps it's called `find`. And in principle, I think you could implement the semantics via separate calls to `insert` and `find` (as long as `supports_concurrent_insert_find()` returns `true`).

**Comment by hahnjo** [comment]
@jrhemstad @PointKernel @sleeepyjack is this good to merge?

**Pull request merged**
